### PR TITLE
Fix bugs with usage flush logic and improve concurrency tests

### PR DESCRIPTION
**Pull request by bduffany**
* Fix the `final_before_usec` update condition; we want to update the usage row only if it is finalized up to the _start_ of the current collection period, not the end
* Improve the concurrent access tests to disable Redlock, so we can test that concurrent flushing is safe even if Redis fails. Also assert that the result is correct rather than just asserting that there was no error (this is what caught the bug above).

---

**Version bump**: Patch

### CODE REVIEW

File: enterprise/server/usage/usage.go [hunk 1]
```diff
@@ -79,16 +78,24 @@ var (
...
 
 type TrackerOpts struct {
+	// TEST ONLY options
+
+	// Clock overrides the clock used by the usage tracker.
 	Clock timeutil.Clock
+
+	// DisableRedlock disables the Redis locking mechanism used to prevent
+	// flush operations from accessing the DB concurrently.
+	DisableRedlock bool
```
Comment (tylerwilliams) [inline enterprise/server/usage/usage.go:88]: Test only options that disable a big chunk of the impl are kinda gross -- could we pass in a distributed lock interface in the constructor and then implement that with a regular old lock or no-ops for testing?

type DistributedLock interface {
  Lock(ctx context.Context, key string, expiry time.Duration)
  Unlock(ctx context.Context, key string)
}

File: enterprise/server/usage/usage.go [hunk 2]
```diff
@@ -254,12 +264,9 @@ func (ut *tracker) flushCounts(ctx context.Context, groupID string, c collection
 		PeriodStartUsec: timeutil.ToUsec(c.UsagePeriod().Start()),
 	}
 	dbh := ut.env.GetDBHandle()
-	return dbh.Transaction(ctx, func(tx *db.DB) error {
```
Comment (tylerwilliams) [inline enterprise/server/usage/usage.go:257]: https://www.percona.com/blog/2014/10/28/how-to-deal-with-mysql-deadlocks/

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
I'm hesitant to just remove all transactions here when we don't understand the reason we're getting a deadlock, even though I think what you're suggesting about the queries being safe is probably true right now?

File: enterprise/server/usage/usage.go (hunk 1 above)
Comment (bduffany) [reply enterprise/server/usage/usage.go]: Done

File: enterprise/server/usage/usage.go (hunk 2 above)
Comment (bduffany) [reply enterprise/server/usage/usage.go:257]: (Working on this)

Comment (bduffany) [reply enterprise/server/usage/usage.go:257]: Decided that we'll keep this in a transaction and that we can tolerate these deadlocks once in a blue moon; the worst that happens is that we just log a warning -- MySQL will ensure that at least one of the contending transactions will go through.

File: enterprise/server/usage/usage.go [hunk 3]
```diff
@@ -79,16 +78,23 @@ var (
...
 
 type TrackerOpts struct {
+	// TEST ONLY options
+
+	// Clock overrides the clock used by the usage tracker.
 	Clock timeutil.Clock
+
+	// FlushLock overrides the distributed lock to serialize access to the DB
+	// across flush jobs.
+	FlushLock interfaces.DistributedLock
```
Comment (tylerwilliams) [inline enterprise/server/usage/usage.go:88]: nit: can you remove this from the opts and make it mandatory (like, accept it as the second param in the NewTracker func?)

File: enterprise/server/usage/usage.go [hunk 4]
```diff
@@ -100,20 +106,16 @@ func NewTracker(env environment.Env, opts *TrackerOpts) (*tracker, error) {
...
-
-	pool := redsync_goredis.NewPool(rdb)
-	rs := redsync.New(pool)
-	flushMutex := rs.NewMutex(
-		redisUsageLockKey,
-		redsync.WithTries(1), redsync.WithExpiry(redisUsageLockExpiry),
-	)
-
+	flushLock := opts.FlushLock
+	if flushLock == nil {
```
Comment (tylerwilliams) [inline enterprise/server/usage/usage.go:110]: nit: this conditional logic is still error-prone. If you accept the lock interface as a param and your impl just always uses it, then we'll be good

File: enterprise/server/usage/usage.go (hunk 2 above)
Comment (tylerwilliams) [reply enterprise/server/usage/usage.go:257]: SG

**Review by tylerwilliams — APPROVED** [review]
LGTM

File: enterprise/server/usage/usage.go (hunk 3 above)
Comment (bduffany) [reply enterprise/server/usage/usage.go]: Done

**Pull request merged**